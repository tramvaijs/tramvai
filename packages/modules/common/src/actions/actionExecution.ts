import flatten from '@tinkoff/utils/array/flatten';
import identity from '@tinkoff/utils/function/identity';
import isPromise from '@tinkoff/utils/is/promise';
import type { Action, ActionParameters, DI_TOKEN } from '@tramvai/core';
import { isTramvaiAction } from '@tramvai/core';
import { ACTION_PARAMETERS } from '@tramvai/core';
import type {
  CONTEXT_TOKEN,
  ActionCondition,
  STORE_TOKEN,
  ActionExecution as Interface,
  EXECUTION_CONTEXT_MANAGER_TOKEN,
  ExecutionContext,
  DEFERRED_ACTIONS_MAP_TOKEN,
} from '@tramvai/tokens-common';
import objectMap from '@tinkoff/utils/object/map';
import type { TramvaiAction, TramvaiActionContext } from '@tramvai/types-actions-state-context';
import { ConditionFailError } from '@tinkoff/errors';
import type { ExtractDependencyType } from '@tinkoff/dippy';
import { ActionChecker } from './actionChecker';
import type { ActionType } from './constants';
import { actionType } from './constants';
import { actionTramvaiReducer } from './actionTramvaiReducer';
import { Deferred } from './deferred/deferred.inline';

const EMPTY_DEPS = {};

export const getParameters = (action: Action): ActionParameters<any, any> =>
  action[ACTION_PARAMETERS];

type ExecutionStatus = 'success' | 'failed' | 'pending' | 'forbidden';

export interface ExecutionState {
  status: ExecutionStatus;
  forbiddenBy?: string;
  state: Record<string, any>;
}

type TransformAction = <T>(action: T) => T;

type AnyActionParameters = ActionParameters<any, any> | TramvaiAction<any, any, any>;

export class ActionExecution implements Interface {
  execution: Map<string, ExecutionState>;
  private actionConditionals: ActionCondition[];
  private context: ExtractDependencyType<typeof CONTEXT_TOKEN>;
  private deferredActionsMap: ExtractDependencyType<typeof DEFERRED_ACTIONS_MAP_TOKEN>;
  private store: ExtractDependencyType<typeof STORE_TOKEN>;
  private executionContextManager: ExtractDependencyType<typeof EXECUTION_CONTEXT_MANAGER_TOKEN>;
  private di: ExtractDependencyType<typeof DI_TOKEN>;
  private transformAction: TransformAction;

  constructor({
    store,
    context,
    deferredActionsMap,
    di,
    executionContextManager,
    actionConditionals,
    transformAction,
  }: {
    actionConditionals: (ActionCondition | ActionCondition[])[] | null;
    store: ExtractDependencyType<typeof STORE_TOKEN>;
    context: ExtractDependencyType<typeof CONTEXT_TOKEN>;
    deferredActionsMap: ExtractDependencyType<typeof DEFERRED_ACTIONS_MAP_TOKEN>;
    di: ExtractDependencyType<typeof DI_TOKEN>;
    executionContextManager: ExtractDependencyType<typeof EXECUTION_CONTEXT_MANAGER_TOKEN>;
    transformAction?: TransformAction;
  }) {
    this.actionConditionals = flatten(actionConditionals ?? []);
    this.context = context;
    this.store = store;
    this.deferredActionsMap = deferredActionsMap;
    this.di = di;
    this.executionContextManager = executionContextManager;
    this.execution = new Map();
    this.transformAction = transformAction || identity;

    const initialState = store.getState(actionTramvaiReducer);

    if (initialState && initialState.serverState) {
      objectMap((value, key) => {
        this.execution.set(key, value);
      }, initialState.serverState);
    }
  }

  async runInContext(
    executionContext: ExecutionContext | null,
    action: Action | TramvaiAction<any, any, any>,
    ...params: any[]
  ): Promise<any> {
    const payload = params[0];
    // TODO: replace type with pure context usage
    const type =
      executionContext?.values.pageActions === true ? actionType.global : actionType.local;
    const parameters = this.getActionParameters(action);

    if (!parameters) {
      throw new Error(
        'Cannot resolve internal data for action. Make sure you are using the result of `declareAction` call as an action'
      );
    }

    const executionState = this.getExecutionState(parameters.name);

    if (!this.canExecuteAction(payload, parameters, executionState, type)) {
      switch (parameters.conditionsFailResult) {
        case 'reject':
          return Promise.reject(
            new ConditionFailError({
              conditionName: executionState.forbiddenBy ?? 'unknown',
              targetName: parameters.name,
            })
          );
        default:
          return Promise.resolve();
      }
    }

    const isDeferredAction = parameters.deferred;

    // will be created when spa run actions mode is "before"
    if (isDeferredAction && !this.deferredActionsMap.get(action.name)) {
      const deferred = new Deferred();
      // avoid unhandled promise rejection
      deferred.promise.catch(() => {});
      this.deferredActionsMap.set(action.name, deferred);
    }

    executionState.status = 'pending';

    return this.executionContextManager.withContext(
      executionContext,
      {
        name: parameters.name,
        values: executionContext?.values.pageActions === true ? { pageActions: false } : undefined,
      },
      (executionActionContext, abortController) => {
        const context = this.createActionContext(
          executionContext,
          executionActionContext,
          abortController,
          parameters
        );

        return Promise.resolve()
          .then(() => {
            if (isTramvaiAction(action)) {
              const result = action.fn.apply(context, params);
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              const deferred = this.deferredActionsMap.get(action.name)!;

              if (isDeferredAction) {
                if (!deferred.isResolved() && !deferred.isRejected() && isPromise(result)) {
                  // eslint-disable-next-line promise/no-nesting
                  result.then(deferred.resolve).catch(deferred.reject);
                }
                // pass success state for deferred actions to the client
                // it will be helpful to detect deferred executions for SPA-transitions only
                // eslint-disable-next-line promise/no-return-wrap
                return Promise.resolve();
              }
              return result;
            }

            return action(this.context, payload, context.deps);
          })
          .then((data) => {
            executionState.status = 'success';
            return data;
          })
          .catch((err) => {
            executionState.status = 'failed';
            throw err;
          });
      }
    );
  }

  async run(action: Action | TramvaiAction<any, any, any>, ...params: any[]): Promise<any> {
    return this.runInContext(null, action, ...params);
  }

  canExecute(action: Action | TramvaiAction<any, any, any>) {
    const parameters = this.getActionParameters(action);
    const executionState = this.getExecutionState(parameters.name);

    // @todo any cases when hadcoded global action type and payload could cause problems?
    return this.canExecuteAction(undefined, parameters, executionState, actionType.global);
  }

  private getActionParameters(action: Action | TramvaiAction<any, any, any>) {
    let parameters: AnyActionParameters;

    // TODO: remove else branch after migration to new declareAction
    if (isTramvaiAction(action)) {
      parameters = action;
    } else {
      this.transformAction(action);
      parameters = getParameters(action);
    }

    return parameters;
  }

  private getExecutionState(name: string) {
    let executionState = this.execution.get(name);
    if (!executionState) {
      executionState = { status: 'pending', state: {} };
      this.execution.set(name, executionState);
    }
    return executionState;
  }

  private canExecuteAction(
    payload: any,
    parameters: ActionParameters<any, any>,
    executionState: ExecutionState,
    type: ActionType
  ) {
    const actionChecker = new ActionChecker(
      this.actionConditionals,
      payload,
      parameters,
      executionState,
      type
    );

    return actionChecker.check();
  }

  private createActionContext(
    parentExecutionContext: ExecutionContext | null,
    actionExecutionContext: ExecutionContext,
    abortController: AbortController,
    parameters: AnyActionParameters
  ): TramvaiActionContext<any> {
    return {
      abortController,
      abortSignal: actionExecutionContext?.abortSignal,
      executeAction: this.runInContext.bind(this, actionExecutionContext),
      deps: parameters.deps ? this.di.getOfDeps(parameters.deps) : EMPTY_DEPS,
      actionType: parentExecutionContext?.values.pageActions ? 'page' : 'standalone',
      dispatch: this.store.dispatch,
      getState: this.store.getState,
    };
  }
}
