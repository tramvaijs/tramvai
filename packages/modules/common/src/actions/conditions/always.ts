import type { ActionCondition } from '@tramvai/tokens-common';

export const alwaysCondition: ActionCondition = {
  key: 'alwaysCondition',
  fn: (checker) => {
    if (checker.conditions.always) {
      checker.allow();
    }
  },
};
