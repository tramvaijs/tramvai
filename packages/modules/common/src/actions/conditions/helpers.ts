export const isServer = typeof window === 'undefined';
export const isBrowser = !isServer;
