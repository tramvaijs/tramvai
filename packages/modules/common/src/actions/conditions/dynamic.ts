import type { ActionCondition } from '@tramvai/tokens-common';
import { isBrowser, isServer } from './helpers';

// @todo make this behaviour default in next tramvai major version?
export const dynamicCondition: ActionCondition = {
  key: 'dynamicCondition',
  fn: (checker) => {
    if (checker.conditions.dynamic) {
      if (isServer) {
        checker.setState({ environment: 'server' });
      }

      if (!checker.getState() || checker.getState().environment === 'browser') {
        checker.allow();
      }

      if (isBrowser) {
        checker.setState({ environment: 'browser' });
      }
    }
  },
};
