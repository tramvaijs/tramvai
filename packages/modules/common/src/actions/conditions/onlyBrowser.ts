import type { ActionCondition } from '@tramvai/tokens-common';
import { isServer } from './helpers';

export const onlyBrowser: ActionCondition = {
  key: 'onlyBrowser',
  fn: (checker) => {
    if (checker.conditions.onlyBrowser && isServer) {
      checker.forbid();
    }
  },
};
