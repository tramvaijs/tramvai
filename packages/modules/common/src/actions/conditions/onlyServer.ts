import type { ActionCondition } from '@tramvai/tokens-common';
import { isBrowser } from './helpers';

export const onlyServer: ActionCondition = {
  key: 'onlyServer',
  fn: (checker) => {
    if (checker.conditions.onlyServer && isBrowser) {
      checker.forbid();
    }
  },
};
