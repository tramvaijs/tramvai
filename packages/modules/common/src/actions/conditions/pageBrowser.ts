import type { ActionCondition } from '@tramvai/tokens-common';
import { isServer } from './helpers';
import { actionType } from '../constants';

export const pageBrowser: ActionCondition = {
  key: 'pageBrowser',
  fn: (checker) => {
    if (checker.conditions.pageBrowser && checker.type === actionType.global && isServer) {
      checker.forbid();
    }
  },
};
