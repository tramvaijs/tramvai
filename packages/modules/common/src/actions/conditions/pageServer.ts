import type { ActionCondition } from '@tramvai/tokens-common';
import { isBrowser } from './helpers';
import { actionType } from '../constants';

export const pageServer: ActionCondition = {
  key: 'pageServer',
  fn: (checker) => {
    if (checker.conditions.pageServer && checker.type === actionType.global && isBrowser) {
      checker.forbid();
    }
  },
};
