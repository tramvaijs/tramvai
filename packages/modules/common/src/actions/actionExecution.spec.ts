import type { TramvaiActionDefinition } from '@tramvai/core';
import { createAction, declareAction, ACTION_PARAMETERS } from '@tramvai/core';
import { createContainer } from '@tinkoff/dippy';
import { ActionExecution } from './actionExecution';
import { alwaysCondition } from './conditions/always';
import { ExecutionContextManager } from '../executionContext/executionContextManager';

const contextMock = { isContext: true };

const delay = (time: number) => new Promise((resolve) => setTimeout(resolve, time));

const legacyActionNameFactory = (name: string, result: string[], conditions = {}) =>
  createAction({
    fn: (c, payload) => result.push(`action: ${name} with: ${payload}`),
    name,
    conditions,
  });

const actionNameFactory = (
  name: string,
  result: string[],
  options: Partial<TramvaiActionDefinition<any, any, any>> = {}
) =>
  declareAction({
    fn(c, payload) {
      return result.push(`action: ${name} with: ${payload}`);
    },
    name,
    ...options,
  });

describe('actionExecution in global page', () => {
  it('Запуск page действий без каких либо опций', async () => {
    const executionContextManager = new ExecutionContextManager();

    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = legacyActionNameFactory('action1', result);
    const action2 = legacyActionNameFactory('action2', result);

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
      }
    );
    expect(result).toMatchInlineSnapshot(`
      [
        "action: action1 with: @globalAction",
        "action: action2 with: @globalAction",
      ]
    `);
    // Экшен не должен выполниться. Автоматическиая дедубликация
    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
        await instance.runInContext(context, action1, '@globalAction');
      }
    );

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action1 with: @globalAction",
        "action: action2 with: @globalAction",
      ]
    `);
  });

  it('Запуск page экшенов с фильтрами', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'neverCondition',
          fn: (cheker) => {
            if (cheker.conditions.never) {
              cheker.forbid();
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(
          context,
          legacyActionNameFactory('action1', result),
          '@globalAction'
        );
        await instance.runInContext(
          context,
          legacyActionNameFactory('action2', result, { never: true }),
          '@globalAction'
        );
        await instance.runInContext(
          context,
          legacyActionNameFactory('action3', result, { never: true }),
          '@globalAction'
        );
        await instance.runInContext(
          context,
          legacyActionNameFactory('action4', result),
          '@globalAction'
        );
      }
    );

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action1 with: @globalAction",
        "action: action4 with: @globalAction",
      ]
    `);
  });

  it('Множественный запуск page actions с проверкой, что только измененые экшены запускаются', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [alwaysCondition],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = legacyActionNameFactory('action multi 1', result);
    const action2 = legacyActionNameFactory('action multi 2', result, { always: true });

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
      }
    );

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action multi 1 with: @globalAction",
        "action: action multi 2 with: @globalAction",
      ]
    `);
    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(
          context,
          legacyActionNameFactory('action multi 1', result),
          '@globalAction'
        );
        await instance.runInContext(context, action2, '@globalAction');
      }
    );

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action multi 1 with: @globalAction",
        "action: action multi 2 with: @globalAction",
        "action: action multi 2 with: @globalAction",
      ]
    `);
  });

  it('Востанавливаем состояние выполненных экшенов и проверяем на влияние будущих экшенов', async () => {
    const state: any = {
      actionTramvai: {
        serverState: {
          'action resume 1': { status: 'success', state: { test: '@globalAction' } },
          'action resume 3': { status: 'success', state: { test: '@globalAction' } },
        },
      },
    };
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return state[reducer.storeName];
          }
          return state;
        },
      } as any,
      actionConditionals: [
        {
          key: 'test',
          fn: (cheker) => {
            const last = cheker.payload;
            if (cheker.getState() !== last) {
              cheker.setState(last);
              cheker.allow();
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = legacyActionNameFactory('action resume 1', result);
    const action2 = legacyActionNameFactory('action resume 2', result);
    const action3 = legacyActionNameFactory('action resume 3', result);

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
        await instance.runInContext(context, action3, '@globalAction');
        await instance.runInContext(context, action3, '@globalAction');
        await instance.runInContext(context, action3, '@globalAction');
      }
    );

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action resume 2 with: @globalAction",
      ]
    `);
  });

  it('Передаем в экшены di, если были добавлены в deps', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer([
        { provide: 'dep1', useValue: 'dep1 value' },
        { provide: 'dep2', useValue: 'dep2 value' },
      ]),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });

    const result: any[] = [];

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(
          context,
          createAction({
            fn: (context, payload, deps) => result.push([context, payload, deps]),
            name: 'action1',
            deps: { dep1: 'dep1', dep2: 'dep2' },
          }),
          '@globalAction'
        );
        await instance.runInContext(
          context,
          createAction({
            fn: (context, payload, deps) => result.push([context, payload, deps]),
            name: 'action2',
            deps: { dep2: 'dep2' },
          }),
          '@globalAction'
        );
        // Экшен не должен выполниться. Автоматическиая дедубликация
        await instance.runInContext(
          context,
          createAction({
            fn: (context, payload, deps) => result.push([context, payload, deps]),
            name: 'action1',
          }),
          '@globalAction'
        );
      }
    );

    expect(result).toMatchSnapshot();
  });

  it('Преобразование экшенов с старого вида в новый через transformAction', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      transformAction: (action: any) => {
        if (!action[ACTION_PARAMETERS]) {
          return createAction({ fn: action, name: action.name });
        }
        return action;
      },
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(
          context,
          legacyActionNameFactory('action1', result),
          '@globalAction'
        );
      }
    );
    const actionOld = (c: any, payload: string) => result.push(payload);

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(context, actionOld as any, '@globalAction');
      }
    );

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action1 with: @globalAction",
        "@globalAction",
      ]
    `);
  });
});

describe('actionExecution in actions', () => {
  it('Запуск множество экшенов без каких либо опций с изначальным стейтом', async () => {
    const executionContextManager = new ExecutionContextManager();
    const state: any = {
      actionTramvai: {
        serverState: {
          action1: { status: 'success', state: {} },
          action2: { status: 'success', state: {} },
        },
      },
    };
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return state[reducer.storeName];
          }
          return state;
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = legacyActionNameFactory('action1', result);
    const action2 = legacyActionNameFactory('action2', result);

    await instance.run(action1, 1);
    await instance.run(action2, 2);

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action1 with: 1",
        "action: action2 with: 2",
      ]
    `);

    await instance.run(action1, 3);
    await instance.run(action1, 4);
    await instance.run(action2, 5);

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action1 with: 1",
        "action: action2 with: 2",
        "action: action1 with: 3",
        "action: action1 with: 4",
        "action: action2 with: 5",
      ]
    `);
  });

  it('Ограничение по ролям', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'roles',
          fn: (cheker) => {
            const { role } = cheker.conditions;
            if (role) {
              if (role !== 'admin') {
                cheker.forbid();
              }
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = legacyActionNameFactory('action 1', result, { role: 'client' });
    const action2 = legacyActionNameFactory('action 2', result, { role: 'admin' });

    await instance.run(action1, 1);
    await instance.run(action2, 2);
    await instance.run(action2, 3);
    await instance.run(action1, 4);
    // Выполняем экшены по одному разу, дальше из-за одинакового conditions, скипаем

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action 2 with: 2",
        "action: action 2 with: 3",
      ]
    `);
  });
});

describe('action conditions', () => {
  it('Получение и использование значений переданных в conditions экшенов', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'value',
          fn: (cheker) => {
            const { role } = cheker.conditions;
            if (role) {
              if (cheker.getState() !== role) {
                cheker.setState(role);
                cheker.allow();
              }
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = legacyActionNameFactory('action 1', result, { role: 'client' });
    const action2 = legacyActionNameFactory('action 2', result, { role: 'admin' });

    await executionContextManager.withContext(
      null,
      { name: 'test', values: { pageActions: true } },
      async (context) => {
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action1, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
        await instance.runInContext(context, action2, '@globalAction');
        await instance.runInContext(context, action1, '@globalAction');
      }
    );
    // Выполняем экшены по одному разу, дальше из-за одинакового conditions, скипаем

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action 1 with: @globalAction",
        "action: action 2 with: @globalAction",
      ]
    `);
  });

  it('Should use conditionsFailResult', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: (reducer?: any) => {
          if (reducer) {
            return undefined;
          }
          return {};
        },
      } as any,
      actionConditionals: [
        {
          key: 'role',
          fn: (checker) => {
            const { role } = checker.conditions;

            if (role !== 'admin') {
              checker.forbid();
            }
          },
        },
      ],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });
    const result: string[] = [];

    const action1 = actionNameFactory('action 1', result, {
      conditions: { role: 'client' },
      conditionsFailResult: 'empty',
    });
    const action2 = actionNameFactory('action 2', result, {
      conditions: { role: 'client' },
      conditionsFailResult: 'reject',
    });
    const action3 = actionNameFactory('action 3', result, {
      conditions: { role: 'admin' },
      conditionsFailResult: 'empty',
    });
    const action4 = actionNameFactory('action 4', result, {
      conditions: { role: 'admin' },
      conditionsFailResult: 'reject',
    });

    await instance.run(action1);
    await expect(instance.run(action2)).rejects.toMatchObject({
      message: 'Condition failed',
      conditionName: 'role',
      targetName: 'action 2',
    });
    await instance.run(action3);
    await instance.run(action4);

    expect(result).toMatchInlineSnapshot(`
      [
        "action: action 3 with: undefined",
        "action: action 4 with: undefined",
      ]
    `);
  });
});

describe('cancelling actions', () => {
  it('should cancel inner actions', async () => {
    const executionContextManager = new ExecutionContextManager();
    const instance = new ActionExecution({
      di: createContainer(),
      store: {
        getState: () => {
          return {};
        },
      } as any,
      actionConditionals: [],
      // @ts-ignore
      context: contextMock,
      executionContextManager,
      deferredActionsMap: new Map(),
    });

    const innerAction1 = declareAction({
      name: 'inner-1',
      async fn() {
        await delay(2000);

        return 'inner-1';
      },
    });
    const innerAction2 = declareAction({
      name: 'inner-2',
      async fn() {
        if (this.abortSignal.aborted) {
          return 'aborted-2';
        }

        return 'inner-2';
      },
    });

    const action = declareAction({
      name: 'root',
      async fn(abort: boolean) {
        abort && this.abortController.abort();

        const promise = this.executeAction(innerAction1).catch(() => 'error-1');

        await Promise.resolve();
        jest.runAllTimers();

        const r1 = await promise;
        const r2 = await this.executeAction(innerAction2).catch(() => 'error-2');

        return `${r1}_${r2}`;
      },
    });

    expect(await instance.run(action, true)).toBe('error-1_error-2');

    expect(await instance.run(action, false)).toBe('inner-1_inner-2');
  });
});
