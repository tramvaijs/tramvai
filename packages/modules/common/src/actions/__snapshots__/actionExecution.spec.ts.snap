// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionExecution in global page Передаем в экшены di, если были добавлены в deps 1`] = `
[
  [
    {
      "isContext": true,
    },
    "@globalAction",
    {
      "dep1": "dep1 value",
      "dep2": "dep2 value",
    },
  ],
  [
    {
      "isContext": true,
    },
    "@globalAction",
    {
      "dep2": "dep2 value",
    },
  ],
]
`;
